.github/workflows/icu4c.yml

```diff
@@ -325,21 +325,20 @@ jobs:
             - '-std=c++20 -stdlib=libc++'
       steps:
         - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
-        - name: Install Clang-16
+        - name: Install Clang-18
           run: |
-            curl -Ls https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
-            sudo apt-add-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-16 main'
-            sudo apt update
-            sudo apt install -qy clang-16 lld-16 libc++-16-dev libc++abi-16-dev
-        - name: Clang-16 build and Test
+            wget https://apt.llvm.org/llvm.sh
```

Comment:
Now when you've found out that there probably never really were any problems with the Ubuntu versions of the compiler, wouldn't it be simpler to just install the compiler directly from the Ubuntu repository, just like all other packages, instead of having this complicated dependency on an external package repository?

As you're rewriting the package installation anyway in this PR, it would seem convenient to do that change instead.

Reply:
We still need to add the external package repository (`sudo apt-add-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-16 main'`), before installing using `apt install`.  Looking at the script [`llvm.sh`](https://apt.llvm.org/llvm.sh), it seems to do the same thing as manual installation (on the left).  

Benefit of using the script, is we don't need to keep track of dependencies, future bug-fixes, or updating to a more modern minor version of clang-18 (much like `ubuntu-latest` for VM-runners) and updating to clang-19 would be a one line change (sort of). 

Since the script comes from official sources (Same as the external repository), security risks for using it should be low.

Reply:
> We still need to add the external package repository (`sudo apt-add-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-16 main'`), before installing using `apt install`.

Why? Why isn't it possible to install directly from the preconfigured Ubuntu repository, like any other package, like the `doxygen` or `python3-pip` packages?

Reply:
I'm not entirely sure... I've been relying on online resources for installation instructions since I don't have a Linux machine available to test this myself. I created a PR to try `sudo apt install -y clang-18`, but it [failed pretty badly](https://github.com/rp9-next/icu/actions/runs/10886318570/job/30206027170?pr=22). Working through VMs via PRs is tricky—would you be able to test the instructions on your machine and share them if you have them handy?

Reply:
> I created a PR to try `sudo apt install -y clang-18`, but it [failed pretty badly](https://github.com/rp9-next/icu/actions/runs/10886318570/job/30206027170?pr=22).

In that error message it says that the system used is Ubuntu "Jammy", so I checked https://packages.ubuntu.com/ and that release doesn't provide any Clang version newer than 15 so it shouldn't be unexpected that version 18 can't be found (and version 15 is too old for us, it still has that known bug).

But this "Jammy" release is now more than 2 years old. Is there anything that would prevent updating our CI to use this year's "Noble" (Ubuntu 24.04 LTS) instead? That one should include Clang 18 in the standard package repository, which would simplify our CI configuration considerably.

> Working through VMs via PRs is tricky—would you be able to test the instructions on your machine and share them if you have them handy?

Is it possible to somehow replicate the CI environment on a local machine? That would be great! How would I do that, do you know where to find instructions?

Reply:
> But this "Jammy" release is now more than 2 years old. Is there anything that would prevent updating our CI to use this year's "Noble" (Ubuntu 24.04 LTS) instead? That one should include Clang 18 in the standard package repository, which would simplify our CI configuration considerably.

We could consider using the Noble VM runner for the Clang 18 job. It's still in [beta](https://github.com/actions/runner-images), but it does come with Clang 18 [preinstalled](https://github.com/actions/runner-images/issues/9848).

However, I think that we generally tend to upgrade to newer versions of clang faster than new ubuntu versions arrive. Clang follows a 6-month release cycle, while Ubuntu LTS versions come every 2 years. For example, Clang 16 was released in April 2023, but Ubuntu 22.04 was not updated to include it at that time.

IMO the script still sounds a better option (Although, it adds an external package repository, Although it adds an external package repository, it's from the official Clang source.)

> Is it possible to somehow replicate the CI environment on a local machine? That would be great! How would I do that, do you know where to find instructions?

I checked, there's this tool [act](https://github.com/nektos/act) which works on docker to run GHA workflows locally using similar CI environment. I will give it a try. (cc: @echeran since he might be interested)

Reply:
> > Working through VMs via PRs is tricky—would you be able to test the instructions on your machine and share them if you have them handy?
> 
> Is it possible to somehow replicate the CI environment on a local machine? That would be great! How would I do that, do you know where to find instructions?

@mihnita and @Manishearth said elsewhere that they have some links/information about working with CI VMs.

Reply:
Not quite, I think there are some hacky ways to ssh into a VM but that's not ideal.

The standard way for dealing with this in GHA is to use the LLVM apt-repository to install precisely the clang version that is needed, or use the script provided by LLVM.

> Why isn't it possible to install directly from the preconfigured Ubuntu repository, like any other package, like the doxygen or python3-pip packages?

Because Debian is super conservative about when it updates things, stable debian tends to have older packages. Python 3 is quite old, _clang_ is old, but clang 16 is only around a year old.

This isn't really a GitHub side issue; this is Debian.

Reply:
Thanks Everyone! Going forward with the script!